### ISSUE TITLE
[BUG] Problems with cuDF at runtime with AGX Xavier

### ISSUE THREAD

**Issue by baldoucam** [rapidsai/cudf#5151]
[log_print_env.txt](https://github.com/rapidsai/cudf/files/4609663/log_print_env.txt)
In my AGX Xavier gives an error when I import cudf from python. It does not offer any error in the compilation phase but it does occur at runtime.

Python 3.6.9 (default, Apr 18 2020, 01:56:04)
[GCC 8.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import rmm
>>> import pyarrow
>>> from numba import cuda
>>> import cudf
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.6/dist-packages/cudf-0.14.0a0+4051.g5f95c9745.dirty-py3.6-linux-aarch64.egg/cudf/__init__.py", line 5, in <module>
    validate_setup(check_dask=False)
  File "/usr/local/lib/python3.6/dist-packages/cudf-0.14.0a0+4051.g5f95c9745.dirty-py3.6-linux-aarch64.egg/cudf/utils/gpu_utils.py", line 10, in validate_setup
    from cudf._cuda.gpu import (
ImportError: /usr/local/lib/python3.6/dist-packages/cudf-0.14.0a0+4051.g5f95c9745.dirty-py3.6-linux-aarch64.egg/cudf/_cuda/gpu.cpython-36m-aarch64-linux-gnu.so: undefined symbol: cudaDriverGetVersion
>>>

I provide the log with the print_env.sh The cuDF installation method is from the source code.

**Comment by harrism** [comment]
Are you able to run any other CUDA programs?

**Comment by baldoucam** [comment]
Yes, I have also tested the CUDA API with a small software and devicequery works fine without problems. In python I have run LSTM with tensorflow on GPU for a study and it also works well. I don't know what could be going on.

**Comment by baldoucam** [comment]
Another thing that is important is that if I use "ldd" with gpu.cpython-36m-aarch64-linux-gnu.so its output is:

linux-vdso.so.1 (0x0000007f88fd8000)
libc.so.6 => /lib/aarch64-linux-gnu/libc.so.6 (0x0000007f88dfe000)
/lib/ld-linux-aarch64.so.1 (0x0000007f88fad000)

The CUDA library is not linked. In this case, if this function is part of the CUDA API, it cannot be found.
But it links well, it does not give any failure when it creates the .so.

**Comment by jdye64** [comment]
@baldoucam I believe this is happening due to the python/cudf/setup.py file not specifying the Xavier CUDA lib64 path which by default is ```/usr/local/cuda/lib64```. If you look [here](https://github.com/rapidsai/cudf/blob/branch-0.14/python/cudf/setup.py#L55) you can see it only uses the python lib dir and the system default lib dir which is not enough. The reason this works in conda is because the python lib path contains the cuda runtime library.

I have made a modification (here)[https://github.com/jdye64/cudf/blob/cudf_xavier/python/cudf/setup.py#L64] and I'm running a build now to test it works. I'm performing a completely fresh build to be sure and that takes a little while but I will keep you updated on those findings.

Also I don't think its just the CUDA RT library that is missing but we will start there first.

**Comment by baldoucam** [comment]
@jdye64 Thanks for the comment. I had a problem with the includes and saw this solution. (https://github.com/rapidsai/cudf/issues/4121) _[BUG] unknown type in template argument, when building cudf python bindings on ARM_ 

With that data I modified the setup.py and I have it this way. I no longer have problems compiling or linking, so I don't understand how it can be giving that error.

My setup.py
...
...
 include_dirs=[
            "/usr/local/lib/dlpack/include/",
            "/usr/local/lib/rmm/include/",
            "/usr/local/lib/include/",
            "/usr/local/lib/cudf",
            "/usr/include/arrow",
            "../../cpp/include/cudf",
            "../../cpp/include",
            "../../cpp/build/include",
            "../../thirdparty/cub",
            "../../thirdparty/libcudacxx/include",
            os.path.dirname(sysconfig.get_path("include")),
            np.get_include(),
            cuda_include_dir,
        ],
        library_dirs=[
            "/usr/local/lib",
            "/usr/local/cuda/lib64",
            "/usr/local/lib/cudf/lib",
            "/usr/local/lib/cudf/",
            get_python_lib(),
            os.path.join(os.sys.prefix, "lib"),
        ],
        libraries=["cudf"],
        language="c++",
        extra_compile_args=["-std=c++14"],
...
..

**Comment by jdye64** [comment]
@baldoucam ahh we still need to add "cudart" to the libraries like

`libraries=["cudf", "cudart"]` in ```library_dirs``` we simply tell the paths but need to invoking the actual linking with ```libraries```

**Comment by baldoucam** [comment]
@jdye64 Ok with the solution for the error:
I_mportError: /usr/local/lib/python3.6/dist-packages/cudf-0.14.0a0+4051.g5f95c9745.dirty-py3.6-linux-aarch64.egg/cudf/_cuda/gpu.cpython-36m-aarch64-linux-gnu.so: undefined symbol: cudaDriverGetVersion_

Now it gives a different error, related to the same library:

ImportError: /home/baldo/cudf/python/cudf/cudf/_cuda/gpu.cpython-36m-aarch64-linux-gnu.so: undefined symbol: cuDeviceGetName

**Comment by baldoucam** [comment]
@jdye64 "ldd" with gpu.cpython-36m-aarch64-linux-gnu.so is now:
        linux-vdso.so.1 (0x0000007facdc4000)
        libcudart.so.10.0 => /usr/local/cuda/lib64/libcudart.so.10.0 (0x0000007faccf9000)
        libc.so.6 => /lib/aarch64-linux-gnu/libc.so.6 (0x0000007facb79000)
        /lib/ld-linux-aarch64.so.1 (0x0000007facd99000)
        libstdc++.so.6 => /usr/lib/aarch64-linux-gnu/libstdc++.so.6 (0x0000007fac9e5000)
        libdl.so.2 => /lib/aarch64-linux-gnu/libdl.so.2 (0x0000007fac9d0000)
        libpthread.so.0 => /lib/aarch64-linux-gnu/libpthread.so.0 (0x0000007fac9a4000)
        librt.so.1 => /lib/aarch64-linux-gnu/librt.so.1 (0x0000007fac98d000)
        libm.so.6 => /lib/aarch64-linux-gnu/libm.so.6 (0x0000007fac8d3000)
        libgcc_s.so.1 => /lib/aarch64-linux-gnu/libgcc_s.so.1 (0x0000007fac8af000)

**Comment by jdye64** [comment]
ok getting closer then. I'm still waiting on my clean build (I shouldn't have done that) once that is done I can mess around more quickly and try to find the other missing ones. I'm willing to bet there are more than a few.

**Comment by baldoucam** [comment]
@jdye64 I solved the previous error by adding "cuda" to the line. My setup.py went like this:  
           ...
            "/usr/local/lib/cudf/",
            "/usr/lib/aarch64-linux-gnu/tegra",
        ],
            libraries=["cudf","cudart","cuda"],
         ...

But a new mistake has come up. When I execute the instruction on the RAPID cover itself, python generates the following error:

>>> import cudf
>>> gdf = cudf.read_csv('/home/baldo/file.csv')
terminate called after throwing an instance of 'thrust::system::system_error'
  what():  parallel_for failed: no kernel image is available for execution on the device
Aborted (core dumped) _

_The file.csv contains:
1,2,3,4
4,5,6,7
8,9,10,11_

In addition I have used "pytest -v" to perform the python test and it also throws errors. Maybe the reason that the previous instruction doesn't work, but it doesn't solve it. The output is as follows:

============= test session starts =====================================
platform linux -- Python 3.6.9, pytest-5.4.2, py-1.8.1, pluggy-0.13.1
rootdir: /home/baldo/cudf/python/cudf
plugins: hypothesis-5.6.0
collecting 4321 items                                                                                                                                                                                             Fatal Python error: Aborted

Current thread 0x0000007f9fecc010 (most recent call first):
  File "/home/baldo/cudf/python/cudf/cudf/core/column/column.py", line 380 in copy
  File "/home/baldo/cudf/python/cudf/cudf/core/index.py", line 1063 in __init__
  File "/home/baldo/cudf/python/cudf/cudf/core/index.py", line 1127 in as_index
  File "/home/baldo/cudf/python/cudf/cudf/core/index.py", line 1141 in as_index
  File "/home/baldo/cudf/python/cudf/cudf/core/series.py", line 133 in __init__
  File "/home/baldo/cudf/python/cudf/cudf/tests/test_dataframe.py", line 4394 in <module>
  File "/home/baldo/.local/lib/python3.6/site-packages/_pytest/assertion/rewrite.py", line 152 in exec_module
  File "<frozen importlib._bootstrap>", line 665 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 955 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 971 in _find_and_load
  File "/home/baldo/.local/lib/python3.6/site-packages/py/_path/local.py", line 701 in pyimport
  File "/home/baldo/.local/lib/python3.6/site-packages/_pytest/python.py", line 511 in _importtestmodule
  File "/home/baldo/.local/lib/python3.6/site-packages/_pytest/python.py", line 443 in _getobj
  File "/home/baldo/.local/lib/python3.6/site-packages/_pytest/python.py", line 261 in obj
  File "/home/baldo/.local/lib/python3.6/site-packages/_pytest/python.py", line 459 in _inject_setup_module_fixture
  File "/home/baldo/.local/lib/python3.6/site-packages/_pytest/python.py", line 446 in collect
  File "/home/baldo/.local/lib/python3.6/site-packages/_pytest/runner.py", line 264 in <lambda>
  File "/home/baldo/.local/lib/python3.6/site-packages/_pytest/runner.py", line 244 in from_call
  File "/home/baldo/.local/lib/python3.6/site-packages/_pytest/runner.py", line 264 in pytest_make_collect_report
  File "/home/baldo/.local/lib/python3.6/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/baldo/.local/lib/python3.6/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/home/baldo/.local/lib/python3.6/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/baldo/.local/lib/python3.6/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/baldo/.local/lib/python3.6/site-packages/_pytest/runner.py", line 382 in collect_one_node
  File "/home/baldo/.local/lib/python3.6/site-packages/_pytest/main.py", line 681 in genitems
  File "/home/baldo/.local/lib/python3.6/site-packages/_pytest/main.py", line 684 in genitems
  File "/home/baldo/.local/lib/python3.6/site-packages/_pytest/main.py", line 490 in _perform_collect
  File "/home/baldo/.local/lib/python3.6/site-packages/_pytest/main.py", line 452 in perform_collect
  File "/home/baldo/.local/lib/python3.6/site-packages/_pytest/main.py", line 257 in pytest_collection
  File "/home/baldo/.local/lib/python3.6/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/baldo/.local/lib/python3.6/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/home/baldo/.local/lib/python3.6/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/baldo/.local/lib/python3.6/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/baldo/.local/lib/python3.6/site-packages/_pytest/main.py", line 246 in _main
  File "/home/baldo/.local/lib/python3.6/site-packages/_pytest/main.py", line 191 in wrap_session
  File "/home/baldo/.local/lib/python3.6/site-packages/_pytest/main.py", line 240 in pytest_cmdline_main
  File "/home/baldo/.local/lib/python3.6/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/baldo/.local/lib/python3.6/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/home/baldo/.local/lib/python3.6/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/baldo/.local/lib/python3.6/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/baldo/.local/lib/python3.6/site-packages/_pytest/config/__init__.py", line 125 in main
  File "/home/baldo/.local/bin/pytest", line 8 in <module>
Aborted (core dumped)

**Comment by baldoucam** [comment]
In another applicationit gave the same runtime error:

_terminate called after throwing an instance of 'thrust::system::system_error'
what(): parallel_for failed: no kernel image is available for execution on the device
Aborted (core dumped)_ 

The reason is when compiling with nvcc and a different computability than the device has.
In this case, when building cuDF from source in the python section I don't know how to specify this step, which may be the reason for this error. 
I can't find any feedback about this topic in different forums I have visited.

**Comment by GregoryKimball** [comment]
The Jetson platform is not something that cuDF supports or tests on, but it could lead to interesting use cases. Please submit a new issue if you run into problems with cuDF on Jetson.
